replay/data_preparator.py

```diff
@@ -186,6 +190,209 @@ def _reindex(self, df: DataFrame, entity: str):
             inv_indexer.setLabels(new_labels)
 
 
+class JoinIndexerMLWriter(MLWriter):
+    """Implements saving the JoinIndexerTransformer instance to disk.
+    Used when saving a trained pipeline.
+    Implements MLWriter.saveImpl(path) method.
+    """
+
+    def __init__(self, instance):
+        super().__init__()
+        self.instance = instance
+
+    def saveImpl(self, path: str) -> None:
+        print(f"Saving {type(self.instance).__name__} to '{path}'")
+
+        spark = SparkSession.getActiveSession()
+
+        init_args = self.instance._init_args
+        sc = spark.sparkContext
+        df = spark.read.json(sc.parallelize([json.dumps(init_args)]))
+        df.coalesce(1).write.mode("overwrite").json(join(path, "init_args.json"))
+
+        self.instance.user_col_2_index_map.write.mode("overwrite").save(join(path, "user_col_2_index_map.parquet"))
+        self.instance.item_col_2_index_map.write.mode("overwrite").save(join(path, "item_col_2_index_map.parquet"))
+
+
+class JoinIndexerMLReader(MLReader):
+    def load(self, path):
+        """Load the ML instance from the input path."""
+
+        spark = SparkSession.getActiveSession()
+        args = spark.read.json(join(path, "init_args.json")).first().asDict(recursive=True)
+        user_col_2_index_map = spark.read.parquet(join(path, "user_col_2_index_map.parquet"))
+        item_col_2_index_map = spark.read.parquet(join(path, "item_col_2_index_map.parquet"))
+
+        indexer = JoinIndexerTransformer(
+            user_col=args["user_col"],
+            user_type=args["user_type"],
+            user_col_2_index_map=user_col_2_index_map,
+            item_col=args["item_col"],
+            item_type=args["item_type"],
+            item_col_2_index_map=item_col_2_index_map,
+
+        )
+
+        return indexer
+
+
+class JoinIndexerTransformer(Transformer, MLWritable, MLReadable):
```

Comment:
Better to unify prefix part of respective estimator and transformer.
For instance, JoinBasedIndexerEstimator and JoinBasedIndexerTransformer.

Reply:
Ok.

It's from old logic. First it was as IndexerEstimator, which produces IndexerTransformer(replay), JoinIndexerTransformer and IterativeIndexerTransformer.